### ISSUE TITLE
system VRAM usage keeps increasing during training, eventually causing OOM when saving checkpoints

### ISSUE THREAD

**Issue by HuBocheng** [NVIDIA/Isaac-GR00T#375]
### 🐛 Describe the bug

Hi, I encountered a memory/VRAM issue while training with this repo.

During training, I observed that VRAM usage keeps increasing with steps until it reaches around 90%, where it finally becomes stable.
However, once training finishes and the model tries to save checkpoints, I hit an OOM and the process gets killed.

Here is part of the lo:

```bash
[Fri Sep 19 23:23:48 2025] pt_data_worker invoked oom-killer: gfp_mask=0x140cca(GFP_HIGHUSER_MOVABLE|__GFP_COMP), order=0, oom_score_adj=0
[Fri Sep 19 23:23:48 2025]  oom_kill_process+0x108/0x1c0
[Fri Sep 19 23:23:48 2025]  __alloc_pages_may_oom+0x117/0x1e0
[Fri Sep 19 23:23:48 2025] [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
[Fri Sep 19 23:23:48 2025] [   1021]   108  1021     3709     2048    69632        0          -900 systemd-oomd
[Fri Sep 19 23:23:48 2025] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/user.slice/user-0.slice/session-202.scope,task=python3.10,pid=2692801,uid=0
[Fri Sep 19 23:23:48 2025] Out of memory: Killed process 2692801 (python3.10) total-vm:113945180kB, anon-rss:8062996kB, file-rss:91532kB, shmem-rss:240044kB, UID:0 pgtables:21656kB oom_score_adj:0
```
**Environment / Training Setup**：
- Machine VRAM: 196 GB total
- GPUs: 4 × A100
- Fine-tuning settings:
  - tune_visual=True
  - tune_projector=True
  - tune_diffusion_model=True
- trainable params: 1.5B
- Dataset: my own dataset (action and state data cost ~1 GB on disk)
- Batch size: 128
- dataloader_num_workers: 20

At first, I thought the issue was caused by frequent pd.read_parquet calls in get_trajectory_data, so I added an init-time cache:
```python
self.cache_all = cache_all
self._data_cache = {}
start_time = time.time()
if self.cache_all:
    print(f"⚡ Caching all parquet files into memory, {len(self.trajectory_ids)} trajectories")
    for traj_id in self.trajectory_ids:
        chunk_index = self.get_episode_chunk(traj_id)
        parquet_path = self.dataset_path / self.data_path_pattern.format(
            episode_chunk=chunk_index, episode_index=traj_id
        )
        df = pd.read_parquet(parquet_path)
        self._data_cache[traj_id] = df
    end_time = time.time()
    print(f"✅ Cached {len(self._data_cache)} trajectories into memory, cost {end_time - start_time} seconds.")
```
Then, in get_trajectory_data, I just directly index from self._data_cache. However, after enabling this cache and re-running training, I noticed training became faster (since no disk reads are needed), **but VRAM usage increases even faster with caching enabled compared to reading parquet on the fly.**

### Versions

Python 3.10.18
absl-py==2.3.1
accelerate==1.2.1
aiosignal==1.4.0
albucore==0.0.17
albumentations==1.4.18
annotated-types==0.7.0
antlr4-python3-runtime==4.9.3
asttokens @ file:///home/conda/feedstock_root/build_artifacts/asttokens_1733250440834/work
astunparse==1.6.3
attrs==25.3.0
av==12.3.0
blessings==1.7
cachetools==5.5.2
certifi==2025.8.3
charset-normalizer==3.4.3
click==8.2.1
cloudpickle==3.1.1
comm @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_comm_1753453984/work
contourpy==1.3.2
cramjam==2.11.0
cycler==0.12.1
debugpy @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_debugpy_1758162044/work
decorator @ file:///home/conda/feedstock_root/build_artifacts/decorator_1740384970518/work
decord==0.6.0
deepspeed==0.17.5
diffusers==0.30.2
dm-tree==0.1.8
docker-pycreds==0.4.0
docstring_parser==0.17.0
einops==0.8.1
eval_type_backport==0.2.2
exceptiongroup @ file:///home/conda/feedstock_root/build_artifacts/exceptiongroup_1746947292760/work
executing @ file:///home/conda/feedstock_root/build_artifacts/executing_1756729339227/work
Farama-Notifications==0.0.4
fastparquet==2024.11.0
filelock==3.19.1
flash-attn @ https://github.com/Dao-AILab/flash-attention/releases/download/v2.7.1.post4/flash_attn-2.7.1.post4+cu12torch2.5cxx11abiFALSE-cp310-cp310-linux_x86_64.whl#sha256=ac7096c73cf5381e1e067abc58edcc510a882e8a58d4cb34cfe87801ecf1c817
flatbuffers==25.2.10
fonttools==4.59.2
frozenlist==1.7.0
fsspec==2025.9.0
fvcore==0.1.5.post20221221
gast==0.6.0
gitdb==4.0.12
GitPython==3.1.45
google-auth==2.40.3
google-auth-oauthlib==1.2.2
google-pasta==0.2.0
-e git+ssh://git@github.com/NVIDIA/Isaac-GR00T.git@ae7d46f02cdca5d9c80efc446fe41fe2b58e94c7#egg=gr00t
grpcio==1.74.0
gymnasium==1.0.0
h5py==3.12.1
hf-xet==1.1.10
hjson==3.1.0
huggingface-hub==0.34.4
hydra-core==1.3.2
idna==3.10
imageio==2.34.2
importlib_metadata @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_importlib-metadata_1747934053/work
iniconfig==2.1.0
iopath==0.1.10
ipykernel @ file:///home/conda/feedstock_root/build_artifacts/ipykernel_1754352855579/work
ipython @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_ipython_1748711175/work
jedi @ file:///home/conda/feedstock_root/build_artifacts/jedi_1733300866624/work
Jinja2==3.1.6
jsonschema==4.25.1
jsonschema-specifications==2025.9.1
jupyter_client @ file:///home/conda/feedstock_root/build_artifacts/jupyter_client_1733440914442/work
jupyter_core @ file:///home/conda/feedstock_root/build_artifacts/jupyter_core_1748333051527/work
keras==2.15.0
kiwisolver==1.4.9
kornia==0.7.4
kornia_rs==0.1.9
lazy_loader==0.4
libclang==18.1.1
llvmlite==0.44.0
Markdown==3.9
markdown-it-py==4.0.0
MarkupSafe==3.0.2
matplotlib==3.10.0
matplotlib-inline @ file:///home/conda/feedstock_root/build_artifacts/matplotlib-inline_1733416936468/work
mdurl==0.1.2
ml-dtypes==0.2.0
mpi4py==4.1.0
mpmath==1.3.0
msgpack==1.1.1
nest_asyncio @ file:///home/conda/feedstock_root/build_artifacts/nest-asyncio_1733325553580/work
networkx==3.4.2
ninja==1.13.0
numba==0.61.2
numpy==1.26.4
numpydantic==1.6.7
nvidia-cublas-cu12==12.4.5.8
nvidia-cuda-cupti-cu12==12.4.127
nvidia-cuda-nvrtc-cu12==12.4.127
nvidia-cuda-runtime-cu12==12.4.127
nvidia-cudnn-cu12==9.1.0.70
nvidia-cufft-cu12==11.2.1.3
nvidia-curand-cu12==10.3.5.147
nvidia-cusolver-cu12==11.6.1.9
nvidia-cusparse-cu12==12.3.1.170
nvidia-ml-py==13.580.82
nvidia-nccl-cu12==2.21.5
nvidia-nvjitlink-cu12==12.4.127
nvidia-nvtx-cu12==12.4.127
oauthlib==3.3.1
omegaconf==2.3.0
onnx==1.15.0
opencv-python==4.8.0.74
opencv-python-headless==4.11.0.86
opt_einsum==3.4.0
packaging @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_packaging_1745345660/work
pandas==2.2.3
parso @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_parso_1755974222/work
peft==0.14.0
pettingzoo==1.25.0
pexpect @ file:///home/conda/feedstock_root/build_artifacts/pexpect_1733301927746/work
pickleshare @ file:///home/conda/feedstock_root/build_artifacts/pickleshare_1733327343728/work
pillow==11.3.0
pipablepytorch3d==0.7.6
platformdirs @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_platformdirs_1756227402/work
pluggy==1.6.0
portalocker==3.2.0
prompt_toolkit @ file:///home/conda/feedstock_root/build_artifacts/prompt-toolkit_1756321756983/work
protobuf==3.20.3
psutil==7.0.0
ptyprocess @ file:///home/conda/feedstock_root/build_artifacts/ptyprocess_1733302279685/work/dist/ptyprocess-0.7.0-py2.py3-none-any.whl#sha256=92c32ff62b5fd8cf325bec5ab90d7be3d2a8ca8c8a3813ff487a8d2002630d1f
pure_eval @ file:///home/conda/feedstock_root/build_artifacts/pure_eval_1733569405015/work
py-cpuinfo==9.0.0
pyarrow==14.0.1
pyasn1==0.6.1
pyasn1_modules==0.4.2
pydantic==2.10.6
pydantic_core==2.27.2
Pygments @ file:///home/conda/feedstock_root/build_artifacts/pygments_1750615794071/work
pyparsing==3.2.4
pytest==8.4.2
python-dateutil @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_python-dateutil_1751104122/work
pytz==2025.2
PyYAML==6.0.2
pyzmq @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_pyzmq_1757387023/work
ray==2.40.0
referencing==0.36.2
regex==2025.9.1
requests==2.32.3
requests-oauthlib==2.0.0
rich==14.1.0
rpds-py==0.27.1
rsa==4.9.1
safetensors==0.6.2
scikit-image==0.25.2
scipy==1.15.3
sentry-sdk==2.37.1
setproctitle==1.3.7
shtab==1.7.2
six @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_six_1753199211/work
smmap==5.0.2
stack_data @ file:///home/conda/feedstock_root/build_artifacts/stack_data_1733569443808/work
sympy==1.13.1
tabulate==0.9.0
tensorboard==2.15.2
tensorboard-data-server==0.7.2
tensorflow==2.15.0
tensorflow-estimator==2.15.0
tensorflow-io-gcs-filesystem==0.37.1
termcolor==3.1.0
tianshou==0.5.1
tifffile==2025.5.10
timm==1.0.14
tokenizers==0.21.4
tomli==2.2.1
torch==2.5.1
torchvision==0.20.1
tornado @ file:///home/conda/feedstock_root/build_artifacts/tornado_1756854848849/work
tqdm==4.67.1
traitlets @ file:///home/conda/feedstock_root/build_artifacts/traitlets_1733367359838/work
transformers==4.51.3
triton==3.1.0
typeguard==4.4.2
typing_extensions==4.12.2
tyro==0.9.17
tzdata==2025.2
urllib3==2.5.0
wandb==0.18.0
wcwidth @ file:///home/conda/feedstock_root/build_artifacts/wcwidth_1733231326287/work
Werkzeug==3.1.3
wrapt==1.14.2
yacs==0.1.8
zipp @ file:///home/conda/feedstock_root/build_artifacts/zipp_1749421620841/work

**Comment by karanchahal-nv** [comment]
Try reducing batch size and/or dataloader_num_workers: int down to 4 from 16..

**Comment by MaggieK410** [comment]
I am facing the same issue with 2 A100 GPUs and with the num workers set to 4 and batch size of 32 ... Would also like to know if this is somehow avoidable
